File: clients/android/app/src/main/kotlin/my/ssdid/drive/domain/repository/AuthRepository.kt
package my.ssdid.drive.domain.repository

import my.ssdid.drive.domain.model.LinkedLogin
import my.ssdid.drive.domain.model.TokenInvitation
import my.ssdid.drive.domain.model.TotpSetupInfo
import my.ssdid.drive.domain.model.User
import my.ssdid.drive.util.Result

/**
 * Repository interface for authentication operations.
 *
 * Authentication methods:
 * - Email + TOTP: Email as identifier, TOTP as proof
 * - OIDC: Google and Microsoft sign-in via native SDKs
 * - Account linking: Multiple login methods per account
 */
interface AuthRepository {

    /**
     * Check if user is authenticated (has valid session token).
     */
    suspend fun isAuthenticated(): Boolean

    /**
     * Get the current session token.
     */
    suspend fun getSession(): String?

    /**
     * Save session tokens (access + refresh).
     */
    suspend fun saveSession(accessToken: String, refreshToken: String)

    /**
     * Logout the current user.
     * Clears all tokens and key material.
     */
    suspend fun logout(): Result<Unit>

    /**
     * Get the current authenticated user.
     */
    suspend fun getCurrentUser(): Result<User>

    /**
     * Update the current user's profile.
     */
    suspend fun updateProfile(displayName: String?): Result<User>

    /**
     * Check if keys are unlocked.
     */
    suspend fun areKeysUnlocked(): Boolean

    // ==================== Email + TOTP Auth ====================

    /**
     * Initiate email login.
     * @return true if TOTP verification is required
     */
    suspend fun emailLogin(email: String): Result<Boolean>

    /**
     * Register via email with invitation token. Sends OTP.
     */
    suspend fun emailRegister(email: String, invitationToken: String): Result<Unit>

    /**
     * Verify email registration OTP and create account.
     */
    suspend fun emailRegisterVerify(email: String, code: String, invitationToken: String): Result<User>

    /**
     * Verify TOTP code for login.
     */
    suspend fun totpVerify(email: String, code: String): Result<User>

    /**
     * Initiate TOTP setup. Returns setup info with otpauth URI.
     */
    suspend fun totpSetup(): Result<TotpSetupInfo>

    /**
     * Confirm TOTP setup with first code. Returns backup codes.
     */
    suspend fun totpSetupConfirm(code: String): Result<List<String>>

    /**
     * Initiate TOTP recovery (sends email OTP).
     */
    suspend fun totpRecovery(email: String): Result<Unit>

    /**
     * Verify TOTP recovery code.
     */
    suspend fun totpRecoveryVerify(email: String, code: String): Result<User>

    // ==================== OIDC Auth ====================

    /**
     * Verify an OIDC ID token for login or registration.
     * @param provider "google" or "microsoft"
     * @param idToken The ID token from the native SDK
     * @param invitationToken Optional invitation token for registration
     */
    suspend fun oidcVerify(provider: String, idToken: String, invitationToken: String? = null): Result<User>

    // ==================== Account Logins (Linking) ====================

    /**
     * List linked logins for the current account.
     */
    suspend fun getLinkedLogins(): Result<List<LinkedLogin>>

    /**
     * Initiate linking an email login (sends OTP).
     */
    suspend fun linkEmail(email: String): Result<Unit>

    /**
     * Verify email link OTP.
     */
    suspend fun linkEmailVerify(email: String, code: String): Result<LinkedLogin>

    /**
     * Link an OIDC login to the current account.
     */
    suspend fun linkOidc(provider: String, idToken: String): Result<LinkedLogin>

    /**
     * Unlink a login method. Must keep at least 1.
     */
    suspend fun unlinkLogin(loginId: String): Result<Unit>

    // ==================== Biometric Unlock ====================

    suspend fun enableBiometricUnlock(): Result<Unit>
    suspend fun disableBiometricUnlock(): Result<Unit>
    suspend fun unlockWithBiometric(): Result<Unit>
    suspend fun isBiometricUnlockEnabled(): Boolean
    suspend fun lockKeys()

    // ==================== Invitation Token (Public - for new users) ====================

    /**
     * Get public invitation info by token.
     */
    suspend fun getInvitationInfo(token: String): Result<TokenInvitation>

    // ==================== Legacy SSDID (kept during migration) ====================

    suspend fun createChallenge(action: String): ChallengeInfo
    suspend fun launchWalletAuth(challenge: ChallengeInfo)
    suspend fun listenForSession(challenge: ChallengeInfo): String
    suspend fun launchWalletInvite(token: String)
}

/**
 * Information about a challenge for SSDID Wallet authentication.
 */
data class ChallengeInfo(
    val challengeId: String,
    val subscriberSecret: String,
    val walletDeepLink: String
)

File: clients/ios/SsdidDrive/SsdidDrive/Domain/Repository/AuthRepository.swift
import Foundation

/// Repository for authentication operations.
/// Authentication is SSDID wallet-based (QR challenge-response).
/// Password is only used for client-side key encryption during invitation acceptance.
protocol AuthRepository: AnyObject {

    // MARK: - Authentication

    /// Logout the current user
    func logout() async throws

    /// Refresh the access token
    func refreshToken() async throws

    /// Get current user info
    func getCurrentUser() async throws -> User

    // MARK: - Authentication State

    /// Check if user is authenticated
    func isAuthenticated() async -> Bool

    /// Get the current user ID
    var currentUserId: String? { get }

    // MARK: - Biometric Authentication

    /// Check if biometric unlock is available on device
    func isBiometricAvailable() -> Bool

    /// Check if biometric unlock is enabled
    func isBiometricUnlockEnabled() async -> Bool

    /// Disable biometric unlock
    func disableBiometricUnlock() async throws

    /// Authenticate using biometric and unlock keys
    func authenticateWithBiometric() async throws -> Bool

    // MARK: - Key Management

    /// Check if keys are currently unlocked
    func areKeysUnlocked() async -> Bool

    /// Lock keys (clear from memory)
    func lockKeys() async

    // MARK: - Device Management

    /// Enroll current device
    func enrollDevice(name: String) async throws -> Device

    /// Get list of user's devices
    func getDevices() async throws -> [Device]

    /// Revoke a device
    func revokeDevice(deviceId: String) async throws

    /// Get current device ID
    var currentDeviceId: String? { get }

    // MARK: - Invitation Token (Public - for new users)

    /// Get public invitation info by token.
    /// This is for new users who received an invitation link.
    /// No authentication required.
    func getInvitationInfo(token: String) async throws -> TokenInvitation

    /// Accept an invitation and register a new account.
    /// Generates key pairs and encrypts them with the password.
    /// No authentication required - this creates the account.
    func acceptInvitation(token: String, displayName: String, password: String) async throws -> InviteUser

    // MARK: - Wallet-Based Invitation

    /// Launch SSDID Wallet to accept an invitation.
    func launchWalletInvite(token: String) async throws

    /// Save session from wallet callback (invitation acceptance).
    func saveSessionFromWallet(sessionToken: String) async throws

    // MARK: - Multi-Auth Invitation Acceptance

    /// Accept an invitation as an already-authenticated user.
    /// No key generation needed — the user already has an account.
    func acceptInvitationAsExistingUser(token: String) async throws

    /// Accept an invitation via OIDC provider authentication.
    /// Registers a new user using OIDC and accepts the invitation in one step.
    func acceptInvitationWithOidc(token: String, provider: String, idToken: String) async throws
}
